### PR TITLE
Unit Test Cases For compute_weighted_sum() added

### PR DESCRIPTION
**Pull request by surbhi498**
Task 
https://github.com/kaizen-ai/kaizenflow/issues/1000#issue-2313666020

### CODE REVIEW

File: helpers/test/test_hpandas.py [hunk 1]
```diff
@@ -3891,25 +3891,19 @@ def get_multiindex_df(
...
-        ```
-                            column1     column2
-        index   timestamp
-        index1  string1     -0.122140   -1.949431
-                string2     1.303778    -0.288235
-        index2  string1     1.237079    1.168012
-                string2     1.333692    1.708455
-        ```
+        ```                     column1     column2 index   timestamp
+        index1  string1     -0.122140   -1.949431         string2
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:3904]: this is a linter stab
revert all the changes, not related to your PR, back to their original state
check yourself with `Files changed` section in the PR

File: helpers/test/test_hpandas.py [hunk 2]
```diff
@@ -4043,3 +4037,175 @@ def test2(self) -> None:
         is_ok=False
         """
         self.assert_equal(actual_exception, expected_exception, fuzzy_match=True)
+
+
+# #############################################################################
+
+class Test_compute_weighted_sum(hunitest.TestCase):
+    def helper1(self) -> Tuple[Dict[str, pd.DataFrame], pd.DataFrame]:
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:4045]: There is no need to create helper functions if their outputs are used in 1 test method

File: helpers/test/test_hpandas.py [hunk 3]
```diff
@@ -4043,3 +4037,175 @@ def test2(self) -> None:
...
+        weights = pd.DataFrame(weights_data, index=idx3)
+        return dfs, weights
+
+    def test_compute_weighted_sum1(self) -> None:
+        """
+        Check that weighted sums are computed correctly for index
+        `mode="assert_equal"`.
+        """
+        # Get test data.
+        dfs, weights = self.helper1()
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:4102]: So instead of calling for helpers and complicating the code reading in every test case just do:
```
# Create test data.
data1 = { ...
index1 = ...
df1 = ...
# ...
data2 = { ...
index2 = ...
df2 = ...
#
dfs = {"df1": df1, "df2": df2}
weights = ...
index_mode = ...
#
weighted_sums = hpandas.compute_weighted_sum( ...
...
```

use `Test_merge_dfs1` for reference

this way you will make test code visible and clear

File: helpers/test/test_hpandas.py [hunk 4]
```diff
@@ -4043,3 +4037,175 @@ def test2(self) -> None:
...
+        }
+        for key, expected_signature in expected_values.items():
+            actual_df = weighted_sums[key]
+            self.check_df_output(
+                actual_df,
+                expected_length,
+                expected_column_names,
+                expected_column_unique_values=None,
+                expected_signature=expected_signature,
+            )
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:4129]: running cycles in tests always a problem since it is very hard to debug
instead check resulting dict signature
```
actual_signature = str(weighted_sums)
expected_signature = r"""
...
"""
```

see how signature approach is implemented in other tests in the same file
```

File: helpers/test/test_hpandas.py [hunk 5]
```diff
@@ -4043,3 +4037,175 @@ def test2(self) -> None:
...
+
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:4176]: also add a test method that will check that an assertion is raised if input is an empty dict `dfs = {}`
see similar tests with asssertion signature checks in the same file

File: helpers/test/test_hpandas.py (hunk 2 above)
Comment (gpsaggese) [reply helpers/test/test_hpandas.py:4045]: In general it's ok to leave code used once in a separate functions. What's not ok is all the copy-paste here.

File: helpers/test/test_hpandas.py [hunk 6]
```diff
@@ -4043,3 +4037,175 @@ def test2(self) -> None:
...
+        """
+        columns = ["A", "B"]
+        idx1 = [0, 1]
+        idx2 = [2, 3]
+        data1 = [[1, 2], [3, 4]]
+        data2 = [[5, 6], [7, 8]]
+        df1 = pd.DataFrame(data1, columns=columns, index=idx1)
+        df2 = pd.DataFrame(data2, columns=columns, index=idx2)
+        dfs = {"df1": df1, "df2": df2}
+        weights_data = {"w1": [0.2, 0.8]}
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:4091]: Too much copy-paste. Everything is the same but weights_data. Let's move that out and pass it to the same function.

File: helpers/test/test_hpandas.py [hunk 7]
```diff
@@ -4043,3 +4037,175 @@ def test2(self) -> None:
...
+                expected_signature=expected_signature,
+            )
+
+    def test_compute_weighted_sum3(self) -> None:
+        """
+        Check that weighted sums are computed correctly for index `mode
+        ="leave_unchanged"`.
+        """
+        # Get test data.
+        dfs, weights = self.helper3()
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:4183]: Let's move the helpers close to where they are used and then you can use

```
# ######...
```

to separate chunks of code.

**Review by gpsaggese — COMMENTED** [review]
Lots of unfactored out code. Let's clean it up

File: helpers/test/test_hpandas.py [hunk 8]
```diff
@@ -4116,3 +4116,123 @@ def test2(self) -> None:
...
+    @staticmethod
+    def helper(index1: List[int], index2: List[int]) -> Dict[str, pd.DataFrame]:
+        data1 = {"A": [1, 2], "B": [3, 4]}
+        df1 = pd.DataFrame(data1, index=index1)
+        data2 = {"A": [5, 6], "B": [7, 8]}
+        df2 = pd.DataFrame(data2, index=index2)
+        dfs = {"df1": df1, "df2": df2}
+        return dfs
+    
+    def test_compute_weighted_sum1(self) -> None:
```
Comment (gpsaggese) [inline helpers/test/test_hpandas.py:4134]: Ok, but it could have been factored out even more, since you could pass index1, index2, ... and factor out the common pieces.

We are not super strict on the DRY-ness in unit tests, but it's important to recognize any opportunity to simplify and refactor the code.

File: helpers/test/test_hpandas.py [hunk 9]
```diff
@@ -4116,3 +4116,123 @@ def test2(self) -> None:
...
         """
         self.assert_equal(actual_exception, expected_exception, fuzzy_match=True)
+
+
+# #############################################################################
+
+class Test_compute_weighted_sum(hunitest.TestCase):
+    
+    @staticmethod
+    def helper(index1: List[int], index2: List[int]) -> Dict[str, pd.DataFrame]:
```
Comment (DanilYachmenev) [inline helpers/test/test_hpandas.py:4126]: Have you run linter?

File: helpers/test/test_hpandas.py (hunk 8 above)
Comment (DanilYachmenev) [reply helpers/test/test_hpandas.py:4134]: @surbhi498 lets exercise it then

so the proposal here is to make a single `helper` function that will basically implement the functionality of all the test methods:
```
def helper(
    index1: List[int],
    index2: List[int],
    weights: List[float],
    index_mode: str,
    expected_signature: str,
) -> None:
    """
    Build inputs and check that function output is correct.
    """
    # Create test data.
    data1 = {"A": [1, 2], "B": [3, 4]}
    df1 = pd.DataFrame(data1, index=index1)
    data2 = {"A": [5, 6], "B": [7, 8]}
    df2 = pd.DataFrame(data2, index=index2)
    dfs = {"df1": df1, "df2": df2}
    #
    weights = pd.DataFrame(weights_data, index=dfs.keys())
    # Run the function.
    weighted_sums = hpandas.compute_weighted_sum(
        dfs=dfs, weights=weights, index_mode=index_mode
    )
    actual_signature = str(weighted_sums)
    self.assert_equal(actual_signature, expected_signature, fuzzy_match=True)
```

And then you set params in each the corresponding test method and run it

File: helpers/test/test_hpandas.py (hunk 9 above)
Comment (surbhi498) [reply helpers/test/test_hpandas.py:4126]: Yes Sir, I have run Linter and reverted those sections not related to my PR that is stabbed by linter. Should I run linter again?

Comment (DanilYachmenev) [reply helpers/test/test_hpandas.py:4126]: good, I'd just expect this line to be fixed by it because it is 81 char long but it is as it is, no need to do anything here

but in general run linter before every review request (so like after each round of making changes)

**Review by DanilYachmenev — COMMENTED** [review]
lgtm
@samarth9008
